diff --git a/dwm.c b/dwm.c
index 1bbb4b3..6dfad66 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1664,6 +1664,15 @@ void
 toggleview(const Arg *arg) {
 	unsigned int newtagset = selmon->tagset[selmon->seltags] ^ (arg->ui & TAGMASK);
 
+	// the first visible client should be the same after we add a new tag
+	// we also want to be sure not to mutate the focus
+	Client *const c = nexttiled(selmon->clients);
+	if (c) {
+		Client * const selected = selmon->sel;
+		pop(c);
+		focus(selected);
+	}
+
 	if(newtagset) {
 		selmon->tagset[selmon->seltags] = newtagset;
 		focus(NULL);
